File: jan-19-max-square-threshold/solution.java
class Solution {
    public int maxSideLength(int[][] mat, int threshold) {
        int m = mat.length;
        int n = mat[0].length;

        int[][] ps = new int[m + 1][n + 1];
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                ps[i][j] = mat[i - 1][j - 1] + ps[i - 1][j] + ps[i][j - 1] - ps[i - 1][j - 1];
            }
        }

        int lo = 0;
        int hi = Math.min(m, n);
        while (lo < hi) {
            int mid = (lo + hi + 1) / 2;
            boolean ok = false;
            for (int i = 0; i + mid <= m && !ok; i++) {
                for (int j = 0; j + mid <= n; j++) {
                    int sum = ps[i + mid][j + mid] - ps[i][j + mid] - ps[i + mid][j] + ps[i][j];
                    if (sum <= threshold) {
                        ok = true;
                        break;
                    }
                }
            }
            if (ok) lo = mid;
            else hi = mid - 1;
        }
        return lo;
    }
}

File: jan-19-max-square-threshold/solution.py
from typing import List


class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m = len(mat)
        n = len(mat[0])

        ps = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(1, m + 1):
            row = mat[i - 1]
            for j in range(1, n + 1):
                ps[i][j] = row[j - 1] + ps[i - 1][j] + ps[i][j - 1] - ps[i - 1][j - 1]

        def square_sum(r: int, c: int, k: int) -> int:
            r2 = r + k
            c2 = c + k
            return ps[r2][c2] - ps[r][c2] - ps[r2][c] + ps[r][c]

        lo, hi = 0, min(m, n)
        while lo < hi:
            mid = (lo + hi + 1) // 2
            ok = False
            for i in range(0, m - mid + 1):
                if ok:
                    break
                for j in range(0, n - mid + 1):
                    if square_sum(i, j, mid) <= threshold:
                        ok = True
                        break
            if ok:
                lo = mid
            else:
                hi = mid - 1
        return lo
